sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala

```diff
@@ -0,0 +1,172 @@
+ * [0, 1].
+ * The parameter pc can be a DoubleType or DoubleType array.
+ *
+ * The operator is bound to the slower sort based aggregation path because the number of elements
+ * and their partial order cannot be determined in advance. Therefore we have to store all the
+ * elements in memory, and that too many elements can cause GC paused and eventually OutOfMemory
+ * Errors.
+ */
+@ExpressionDescription(
+  usage = """_FUNC_(epxr, pc) - Returns the percentile(s) of expr at pc (range: [0,1]). pc can be
+  a double or double array.""")
+case class Percentile(
+  child: Expression,
+  pc: Expression,
+  mutableAggBufferOffset: Int = 0,
+  inputAggBufferOffset: Int = 0) extends ImperativeAggregate {
+
+  def this(child: Expression, pc: Expression) = {
+    this(child = child, pc = pc, mutableAggBufferOffset = 0, inputAggBufferOffset = 0)
+  }
+
+  private val percentiles: Seq[Double] = pc match {
+    case Literal(ar: GenericArrayData, _: ArrayType) =>
+      ar.asInstanceOf[GenericArrayData].array.map{ d => d.asInstanceOf[Double]}
+    case _ => Seq.empty
+  }
+
+  override def prettyName: String = "percentile"
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ImperativeAggregate =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ImperativeAggregate =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  private var counts = new OpenHashMap[Double, Long]()
+
+  override def children: Seq[Expression] = child :: pc :: Nil
+
+  override def nullable: Boolean = false
+
+  override def dataType: DataType = ArrayType(DoubleType)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, NumericType)
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForOrderingExpr(child.dataType, "function percentile")
+
+  override def supportsPartial: Boolean = false
+
+  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
+
+  override val aggBufferAttributes: Seq[AttributeReference] = percentiles.map(percentile =>
+    AttributeReference(percentile.toString, DoubleType)())
+
+  override val inputAggBufferAttributes: Seq[AttributeReference] =
+    aggBufferAttributes.map(_.newInstance())
+
+  override def initialize(buffer: MutableRow): Unit = {
+    var i = 0
+    while (i < percentiles.size) {
+      buffer.setNullAt(mutableAggBufferOffset + i)
+      i += 1
+    }
+  }
+
+  override def update(buffer: MutableRow, input: InternalRow): Unit = {
+    val v = child.eval(input)
+
+    val key = v match {
+      case o: Byte => o.toDouble
+      case o: Short => o.toDouble
+      case o: Int => o.toDouble
+      case o: Long => o.toDouble
+      case o: Float => o.toDouble
+      case o: Decimal => o.toDouble
+      case o: Double => o
+      case _ => sys.error("Percentile is restricted to Numeric types only.")
```

Comment:
I don't this this is possible. Spark will try to provide numeric types (and inject casts if needed), because you have defined `override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, NumericType)`